staging/src/k8s.io/csi-api/pkg/crd/crd.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crd
+
+import (
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	csiapiv1alpha1 "k8s.io/csi-api/pkg/apis/csi/v1alpha1"
+	"reflect"
+)
+
+// CSIDriverCRD returns the CustomResourceDefinition for CSIDriver object.
+func CSIDriverCRD() *apiextensionsv1beta1.CustomResourceDefinition {
+	return &apiextensionsv1beta1.CustomResourceDefinition{
```

Comment:
For the record, this is not what I had in mind. I was saying to write our usual "Go IDL" ("types.go" files) for the kinds.

My backup preference (if it's difficult to automatically generate yaml CRD output from types.go style definitions) is to just check in the yaml file and make that canonical.

It doesn't seem useful to me to have this go code here, it seems like a really round-about way for keeping the yaml file up to date. Why not delete this and just use only the yaml file?

Reply:
@liggitt 

Long term it'd be ideal to have CRD autogenerated from types.go files. This is only here as a temporary measure for this release and is a model used by RBAC rules as well: https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy_test.go#L162

I haven't had a chance to look into how to use a checked-in YAML file in csi-api from tests, but somehow I have to get the test binary to recognize the appropriate path (presumably by following the test manifest model). Unsure how much work that'll take atm

Reply:
This ensures the yaml matches the in-memory type. The alternative is to start with canonical yaml, but then there's no good way to round trip and ensure zero diff in a test (because the null fields you commented about would show up again when serialized)

Reply:
> It doesn't seem useful to me to have this go code here, it seems like a really round-about way for keeping the yaml file up to date. Why not delete this and just use only the yaml file?

In the places we need in-memory definitions, you'd then need code to load the yaml, and tests to ensure what we loaded matched the on-disk manifests. Seems about equal in complexity to me, and this way seems to allow stronger validation the two are in sync

Reply:
Did some investigation on the YAML-only approach. The test binary relies on go-bindata in order to read a file anywhere. To do that for our CRD YAML file in csi-api, we'll need to add a brand new output path in [hack/generate-bindata.sh](https://github.com/kubernetes/kubernetes/blob/f4c4ae18375912f2bee240ef75de4db2302d3ca9/hack/generate-bindata.sh#L67), and update the appropriate BUILD file to autogenerate bindata. Not too bad, unless there are surprises :)

Reply:
> Why would you need to round trip something stored as yaml? You load it from a file, what is there to test? What failure mode do you want to guard against?

I'd want to protect against a typo or misindent in the yaml

Reply:
You could be looking for syntactical correctness, in which case schema validation should be sufficient.

Or you could be looking for the equivalent of unit tests for k8s objects. Schema validation won't work for that, but neither does round tripping. The e2e test will work for this, I think.

I guess I think that having the e2e test install the yaml and then use it is a stronger test than round-tripping would be, I think that's sufficient for the moment, and the "cheap" submission time check should be a schema validation rather than a round-trip and compare.

Comparing objects is fraught with danger, e.g. the null fields.

Reply:
I'm expecting potential reviewer concerns around adding a new go-bindata directory, as our directory is going to be for a specific purpose. The other two are `test/e2e/generated/bindata.go`, for general test data, and `pkg/generated/bindata.go`, for runtime data. Since the API object definitions are in staging/src/k8s.io/csi-api, it doesn't make sense to have the file in `pkg/`. We could either add it for `staging/`, or `staging/src/k8s.io/csi-api`.

Alternatively, if we were to go with the YAML route, I think it's much better to keep the file in the existing `test_manifest` directory (linked to `test/e2e/generated/bindata.go`). Other than documentation, the only purpose for this file right now is e2e testing. For documentation we could link to this file. The user doesn't care where the file is stored, as long as they have a link. The link doesn't need to change between now and the next release, when we remove the need for users to install CRD manually. Many other tests follow this pattern so we know it works.

Reply:
If you can put a TODO in the go file that makes it clear that the mechanism is temporary and shouldn't be copied by others, then I'll stop objecting.